File: src/env.ts
/**
 * The **only** module in this package that reads the environment.
 *
 * `docs/implementation.md` §2 (MECHANISM PIN — instance state, never globals)
 * pins exactly one environment-reading module per wrapper; everything else
 * receives resolved configuration as a value. That is not a style preference:
 * it is what keeps `Deno.env` / `process.env` out of the operation modules, so
 * no operation can silently acquire a dependency on ambient state and every
 * env-var behaviour has exactly one seam to exercise in tests (§9).
 *
 * The spec names that module `src/config.ts`; this package splits the *reading*
 * (here) from the *resolving* (`src/config.ts`, which imports this file and is
 * the only caller). The invariant the pin is about — one place touches the
 * environment — is unchanged, and it is enforced mechanically: the repo gate
 * greps the whole `src/` tree for `Deno.env|process.env` and requires exactly
 * one matching file.
 *
 * The read is a **capability probe**, not a runtime assumption, so the same
 * source runs under Deno, Node and Bun (`docs/implementation.md` §1: the client
 * core is runtime-neutral TypeScript over Web-standard globals). Where neither
 * runtime object is present the value degrades to "unset" rather than throwing —
 * a browser bundle of this package must still construct a client from explicit
 * arguments.
 *
 * @module
 */

/** Environment variable holding the server **origin** (see `src/config.ts`). */
export const ENV_BASE_URL = "W6W_BASE_URL";

/** Environment variable holding the bearer token. */
export const ENV_TOKEN = "W6W_TOKEN";

/**
 * The shape this module probes for. Declared structurally rather than with
 * `any` (which `deno lint`'s recommended rules reject) so the probe stays
 * typed: every hop is optional, because the whole point is that none of it may
 * exist.
 */
interface EnvCapableGlobal {
  Deno?: { env?: { get?: (name: string) => string | undefined } };
  process?: { env?: Record<string, string | undefined> };
}

/**
 * Read one environment variable, or `undefined` when it carries no usable
 * value.
 *
 * **A variable that is set but empty — or whitespace-only — reads as ABSENT.**
 * `W6W_BASE_URL=`, `W6W_BASE_URL="   "` and an unset `W6W_BASE_URL` are all the
 * same thing here, and the precedence chain falls through all three to the next
 * source (and ultimately to a configuration error).
 *
 * This deviates deliberately from the pinned snippet in
 * `docs/implementation.md` §2, which reads `g.Deno.env.get(name) ?? undefined`.
 * A nullish coalesce does **not** convert `""` to `undefined`, so an empty
 * `W6W_BASE_URL` would survive as a value, reach the join rule, and resolve to
 * the relative URL `"/api"` — exactly the browser same-origin default that §2
 * forbids a library from having. A relative base URL fails later, somewhere
 * else, with a message about the request rather than about the configuration.
 *
 * So the check below is a **truthiness/trim** check and not a nullish one, on
 * purpose. Do not "tidy" it back to `??`. An empty-string environment variable
 * is not a hypothetical: a Docker `ENV`/`--build-arg` that was never given a
 * value produces exactly this, and it looks absent in every log line while
 * behaving as present in every `??`.
 *
 * The value is returned verbatim (not trimmed) — normalising the *shape* of a
 * base URL belongs to `src/config.ts`, which owns the join rule.
 *
 * @param name - The variable to read, e.g. {@linkcode ENV_BASE_URL}.
 * @returns The value, or `undefined` if unset, empty, blank or unavailable.
 */
export function readEnv(name: string): string | undefined {
  const g = globalThis as unknown as EnvCapableGlobal;
  let value: string | undefined;
  if (g.Deno?.env?.get) value = g.Deno.env.get(name);
  else if (g.process?.env) value = g.process.env[name];
  else return undefined;
  if (value === undefined || value.trim().length === 0) return undefined;
  return value;
}

File: src/w6w/_env.py
"""The **only** module in this package that reads the environment.

`docs/implementation.md` §2 (MECHANISM PIN — instance state, never globals) pins
exactly one environment-reading module per wrapper; everything else receives
resolved configuration as a plain value. That is not a style preference. It is
what keeps `os.environ` out of the operation modules, so no operation can
silently acquire a dependency on ambient process state, and so every env-var
behaviour has exactly **one** seam to exercise in tests (§9).

The spec names that module `w6w/_config.py`; this package splits the *reading*
(here) from the *resolving* (`_config.py`, which imports this file and is its
only caller). The invariant the pin is about — one place touches the
environment — is unchanged, and it is enforced mechanically: the repo gate greps
the whole `src/w6w` tree for `os.environ|os.getenv` and requires exactly one
matching file. The `node` wrapper made the same split for the same reason
(`src/env.ts` beside `src/config.ts`).

The mapping is **injectable** (`environ=`) so the whole precedence chain can be
tested without mutating the process environment: a module-patched global cannot
be exercised two different ways in one process, which is the same argument that
makes the credential and the transport constructor-injected.
"""

from __future__ import annotations

import os
from typing import Mapping, Optional

#: Environment variable holding the server **origin** (see `_config.py`).
ENV_BASE_URL: str = "W6W_BASE_URL"

#: Environment variable holding the bearer token.
ENV_TOKEN: str = "W6W_TOKEN"


def process_environ() -> Mapping[str, str]:
    """Return the live process environment.

    Factored out as a function so `read_env`'s default source is a named,
    assertable thing rather than a value captured at import time — and so this
    file stays the single grep-able point where `os.environ` appears.

    :returns: `os.environ` itself (not a copy — reads must see later changes).
    """
    return os.environ


def read_env(name: str, environ: Optional[Mapping[str, str]] = None) -> Optional[str]:
    """Read one environment variable, or `None` when it carries no usable value.

    **A variable that is set but empty — or whitespace-only — reads as ABSENT.**
    `W6W_BASE_URL=`, `W6W_BASE_URL="   "`, `W6W_BASE_URL=$'\\n'` and an unset
    `W6W_BASE_URL` are all the same thing here, and the precedence chain falls
    through all four to the next source (and ultimately to a `ConfigError`).

    The guard is therefore a **truthiness/strip** check and deliberately *not*
    an `is None` test. `os.environ.get` returns `""` for a set-but-empty
    variable, never `None`, so `is None` is the Python spelling of the same
    nullish-coalescing mistake that `docs/implementation.md` §2 forbids in the
    TypeScript wrappers: an empty `W6W_BASE_URL` would survive as a value and
    short-circuit the precedence chain, leaving the client on an empty base and
    issuing every request against a relative URL — exactly the browser
    same-origin default a library must not have. Do not "tidy" this into
    `if value is None`.

    An empty environment variable is not a hypothetical. `export W6W_BASE_URL=`,
    a Dockerfile `ENV W6W_BASE_URL=` with no build arg, and a CI variable whose
    template did not interpolate all produce exactly this, and it *looks* absent
    in every log line while behaving as present in every null check.

    The value is returned **verbatim**, not stripped: trimming applies to the
    emptiness *test*, not to the stored value. Normalising the shape of a base
    URL is the join rule's job (`_config.join_base_url`).

    :param name: The variable to read, e.g. :data:`ENV_BASE_URL`.
    :param environ: Mapping to read from. Defaults to the process environment;
        pass one to test the precedence chain without touching `os.environ`.
    :returns: The value, or `None` if unset, empty or whitespace-only.
    """
    source = process_environ() if environ is None else environ
    value = source.get(name)
    if value is None or not value.strip():
        return None
    return value
